File: leetcode/two pointer/159. Longest Substring with At Most Two Distinct Characters/js/main.js
// Time O(N)
// Space O(K)
const lengthOfLongestSubstringTwoDistinct = str => {
  let map = {};

  for (let x of str) {
    map[x] = 0;
  }

  let start = 0;
  let end = 0;
  let cnt = 0;
  let maxLen = 0;

  while (end < str.length) {
    if (map[str[end]] === 0) {
      cnt++;
    }

    map[str[end]]++;
    end++;

    while (cnt > 2) {
      if (map[str[start]] === 1) {
        cnt--;
      }
      map[str[start]]--;
      start++;
    }

    maxLen = Math.max(maxLen, end - start);
  }

  return maxLen;
};

File: leetcode/two pointer/159. Longest Substring with At Most Two Distinct Characters/py/main.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        mapping = {}
        k = 0
        start, end = 0, 0
        n = len(s)
        startIndex = 0
        maxLen = 0
        
        for ch in s:
            mapping[ch] = 0

        while end < n:
            if mapping[s[end]] == 0:
                k += 1
                
            mapping[s[end]] += 1    
            end += 1

            while k >= 3:
                if mapping[s[start]] == 1:
                    k -= 1

                mapping[s[start]] -= 1
                start += 1
            
            maxLen = max(maxLen, end - start)
        
        return maxLen